src/pages/invoices/common/hooks/useProductColumns.ts

```diff
@@ -16,47 +16,133 @@ export function useProductColumns() {
-        taxes.push('$product.tax_rate1');
-      }
+    const numberOfPdfVariables = pdfVariables.length;
 
-      if (enabledTaxRates > 1) {
-        taxes.push('$product.tax_rate2');
-      }
+    if (!updatedVariables.includes('$product.item')) {
+      updatedVariables.splice(0, 0, '$product.item');
+      pdfVariables.splice(0, 0, '$product.item');
+    }
 
-      if (enabledTaxRates > 2) {
-        taxes.push('$product.tax_rate3');
-      }
+    if (!numberOfPdfVariables) {
+      updatedVariables.push('$product.description');
+    }
+
+    if (!updatedVariables.includes('$product.unit_cost')) {
+      updatedVariables.push('$product.unit_cost');
+      pdfVariables.push('$product.unit_cost');
+    }
+
+    if (!updatedVariables.includes('$product.quantity')) {
+      updatedVariables.push('$product.quantity');
+      pdfVariables.push('$product.quantity');
+    }
+
+    const taxes: string[] = [];
+    const enabledTaxRates = company?.enabled_item_tax_rates || 0;
+
+    if (enabledTaxRates > 0) {
+      taxes.push('$product.tax_rate1');
+    }
+
+    if (enabledTaxRates > 1) {
+      taxes.push('$product.tax_rate2');
+    }
 
-      // Let's remove original tax field because we don't need it anymore,
-      // but first we gonna keep the index, because that's where we are injecting other input fields.
-      const taxVariableIndex = variables.findIndex(
-        (variable) => variable === '$product.tax'
+    if (enabledTaxRates > 2) {
+      taxes.push('$product.tax_rate3');
+    }
+
+    const taxVariableIndex = getColumnIndex(
+      '$product.tax',
+      pdfVariables,
+      updatedVariables
+    );
+
+    updatedVariables.splice(taxVariableIndex, 0, ...taxes);
+
+    pdfVariables.splice(taxVariableIndex, 0, ...taxes);
+
+    updatedVariables = updatedVariables.filter(
+      (variable) => variable !== '$product.tax'
+    );
+
+    pdfVariables = pdfVariables.filter(
+      (variable) => variable !== '$product.tax'
+    );
+
+    if (!company.enable_product_discount) {
+      updatedVariables = updatedVariables.filter(
+        (variable) => variable !== '$product.discount'
+      );
+
+      pdfVariables = pdfVariables.filter(
+        (variable) => variable !== '$product.discount'
       );
+    }
 
-      variables.splice(taxVariableIndex + 1, 0, ...taxes);
+    if (
+      company.enable_product_discount &&
+      !updatedVariables.includes('$product.discount')
+    ) {
+      updatedVariables.push('$product.discount');
+      pdfVariables.push('$product.discount');
+    }
 
-      if (!company.enable_product_discount) {
-        variables = variables.filter(
-          (variable) => variable !== '$product.discount'
+    ['product1', 'product2', 'product3', 'product4'].forEach((field) => {
+      if (
+        company?.custom_fields[field] &&
+        !pdfVariables.includes(`$product.${field}`)
+      ) {
+        updatedVariables.splice(
+          updatedVariables.length,
+          0,
+          `$product.${field}`
```

Comment:
The same goes for the previous one, `splice` is very hard to read and understand at first. It has a terrible name and usage as well, so don't use it unless it's last resort.

Reply:
Yeah, improved!